org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/job/BackendProgressJobScheduler.java

```diff
@@ -0,0 +1,134 @@
+import org.eclipse.core.runtime.Status;
+import org.eclipse.core.runtime.jobs.Job;
+import org.sonarsource.sonarlint.core.clientapi.client.progress.ProgressUpdateNotification;
+import org.sonarsource.sonarlint.core.clientapi.client.progress.StartProgressParams;
+import org.sonarsource.sonarlint.core.commons.log.SonarLintLogger;
+
+/**
+ *  As the backend is now in charge of synchronization we want the user to see the progress even though it was not
+ *  scheduled on the IDE side. Therefore we create fake jobs for every backend job which offers interaction
+ *  possibilities for the user.
+ */
+public class BackendProgressJobScheduler {
+  private ConcurrentHashMap<String, BackendProgressJob> jobPool = new ConcurrentHashMap<>();
+  private static final BackendProgressJobScheduler INSTANCE = new BackendProgressJobScheduler();
+  
+  private BackendProgressJobScheduler() {
+  }
+  
+  public static BackendProgressJobScheduler get() {
+    return INSTANCE;
+  }
+  
+  /** Start a new progress bar by using an IDE job */
+  public CompletableFuture<Void> startProgress(StartProgressParams params) {
+    var taskId = params.getTaskId();
+    if (jobPool.containsKey(taskId)) {
+      var errorMessage = "";
+      SonarLintLogger.get().error(errorMessage);
+      return CompletableFuture.failedFuture(new IllegalArgumentException(errorMessage));
+    }
+    
+    return null;
+  }
+  
+  /** Update the progress bar IDE job */
+  public void update(String taskId, ProgressUpdateNotification notification) {
+    if (!jobPool.containsKey(taskId)) {
+      SonarLintLogger.get().error("");
+      return;
+    }
+    jobPool.get(taskId).update(notification);
+  }
+  
+  /** Complete the progress bar IDE job */
+  public void complete(String taskId) {
+    if (!jobPool.containsKey(taskId)) {
+      SonarLintLogger.get().error("");
+      return;
+    }
+    jobPool.get(taskId).complete();
+  }
+  
+  /** This job is only an IDE frontend for a job running in the SonarLintBackend */
+  private static class BackendProgressJob extends Job {
+    private Object waitMonitor = new Object();
+    private AtomicReference<String> message;
+    private AtomicInteger percentage = new AtomicInteger(0);
+    private AtomicBoolean complete = new AtomicBoolean(false);
+    
+    public BackendProgressJob(StartProgressParams params) {
+      super(params.getTitle());
+      setPriority(DECORATE);
+      
+      this.message = new AtomicReference<>(params.getMessage());
+    }
+
+    @Override
+    protected IStatus run(IProgressMonitor monitor) {
+      try {
+        monitor.setTaskName(message.get());
+        monitor.worked(percentage.get());
+        
+        while (!complete.get()) {
+          synchronized(waitMonitor) {
+            monitor.setTaskName(message.get());
+            monitor.worked(percentage.get());
+          }
+        }
+        
+        monitor.done();
+        return monitor.isCanceled() ? Status.CANCEL_STATUS : Status.OK_STATUS;
```

Comment:
We don't support cancellation at the moment, so I am not sure we should check `isCanceled`. Not a big deal, though.